### PR TITLE
Cp shutdown

### PR DESCRIPTION
**Pull request by willmcgugan**
- Changed binding for command palette to ctrl+backslash
- Renamed `Source.post_init` to `Source.startup`
- Added a `Source.shutdown` method.
- Renamed `command.Source` to `command.Provider`

### CODE REVIEW

File: src/textual/command.py [hunk 1]
```diff
@@ -101,7 +102,7 @@ def __post_init__(self) -> None:
 """Return type for the command source match searching method."""
 
 
-class Source(ABC):
+class Provider(ABC):
     """Base class for command palette command sources.
```
Comment (davep) [inline src/textual/command.py:106]:
```suggestion
    """Base class for command palette command providers.
```

File: src/textual/command.py [hunk 2]
```diff
@@ -205,7 +206,15 @@ async def search(self, query: str) -> Hits:
         """
         yield NotImplemented
 
+    async def shutdown(self) -> None:
+        """Called when the Source is shutdown.
```
Comment (davep) [inline src/textual/command.py:210]:
```suggestion
        """Called when the Provider is shutdown.
```

**Comment by willmcgugan** [comment]
Patience please @davep

**Comment by willmcgugan** [comment]
OK, knock yourself out.

File: docs/examples/guide/command_palette/command01.py [hunk 3]
```diff
@@ -3,19 +3,19 @@
...
 from textual.app import App, ComposeResult
-from textual.command import Hit, Hits, Source
+from textual.command import Hit, Hits, Provider
 from textual.containers import VerticalScroll
 from textual.widgets import Static
 
 
-class PythonFileSource(Source):
+class PythonFileCommands(Provider):
     """A command source to open a Python file in the current working directory."""
```
Comment (davep) [inline docs/examples/guide/command_palette/command01.py]:
```suggestion
    """A command provider to open a Python file in the current working directory."""
```

File: src/textual/screen.py [hunk 4]
```diff
@@ -157,10 +157,10 @@ class Screen(Generic[ScreenResultType], Widget):
     title: Reactive[str | None] = Reactive(None, compute=False)
     """Screen title to override [the app title][textual.app.App.title]."""
 
-    COMMAND_SOURCES: ClassVar[set[type[Source]]] = set()
-    """Command sources used by the [command palette](/guide/command), associated with the screen.
+    COMMANDS: ClassVar[set[type[Provider]]] = set()
+    """Command providers used by the [command palette](/guide/command), associated with the screen.
 
-    Should be a set of [command.Source][textual.command.Source] classes.
+    Should be a set of [command.Provider][textual.command.Provider] classes.
```
Comment (davep) [inline src/textual/screen.py:163]:
```suggestion
    Should be a set of [`command.Provider`][textual.command.Provider] classes.
```

File: src/textual/command.py [hunk 5]
```diff
@@ -504,11 +513,21 @@ def on_mount(self, _: Mount) -> None:
...
+        self._providers = [
+            provider_class(self._calling_screen, match_style)
+            for provider_class in self._provider_classes
         ]
-        for _source in self._sources:
-            _source._post_init()
+        for provider in self._providers:
+            provider._post_init()
+
+    async def on_unmount(self) -> None:
```
Comment (davep) [inline src/textual/command.py:523]: There's an asymmetry with how exceptions in `shutdown` are handled, vs `startup` (the latter reporting as errors in `search` does too). An error in `startup` is nicely reported in the console:

Screenshot 2023-09-14 at 11 32 18

whereas an error in `showdown` is like this:

Screenshot 2023-09-14 at 11 32 35

File: src/textual/widgets/_header.py [hunk 6]
```diff
@@ -22,11 +22,19 @@ class HeaderIcon(Widget):
...
+
+    HeaderIcon:hover {
+        background: $foreground 10%;
+    }
     """
 
     icon = Reactive("⭘")
     """The character to use as the icon within the header."""
 
+    async def on_click(self, event: Click) -> None:
```
Comment (davep) [inline src/textual/widgets/_header.py:34]: Deserves a docstring to state its purpose here.

**Review by davep — CHANGES_REQUESTED** [review]
Looking good; liking the icon thing.

However, I think there might be an issue with handling errors in the new `shutdown` method.

**Review by davep — APPROVED** [review]
Noticed and squished a couple of remaining source->provider rewordings.

Looks good to me. SHIP IT!

**Pull request merged**